plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+insert (P.toBuiltinData -> k) (P.toBuiltinData -> a) m =
+  unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        (BI.mkCons (BI.mkPairData k a) nil)
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then BI.mkCons (BI.mkPairData k a) tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE delete #-}
+delete :: forall k a. (P.ToData k) => k -> Map k a -> Map k a
+delete (P.toBuiltinData -> k) m = unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        nil
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE singleton #-}
+singleton :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a
+singleton (P.toBuiltinData -> k) (P.toBuiltinData -> a) = unsafeFromBuiltinList xs
+  where
+    xs = BI.mkCons (BI.mkPairData k a) nil
+
+{-# INLINEABLE empty #-}
+empty :: forall k a. Map k a
+empty = unsafeFromBuiltinList nil
+
+{-# INLINEABLE null #-}
+null :: forall k a. Map k a -> Bool
+null = P.null . toBuiltinList
+
+{-# INLINEABLE unsafeFromList #-}
+unsafeFromList :: (P.ToData k, P.ToData a) => [(k, a)] -> Map k a
+unsafeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+
+{-# INLINEABLE uncons #-}
+uncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  Maybe ((k, a), Map k a)
+uncons m = case P.uncons (toBuiltinList m) of
+  Nothing -> Nothing
+  Just (pair, rest) ->
+    let (k, a) = P.pairToPair pair
+     in Just ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList rest)
+
+{-# INLINEABLE unsafeUncons #-}
+unsafeUncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  ((k, a), Map k a)
+unsafeUncons m =
+  ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList tl)
+  where
+    (hd, tl) = P.unsafeUncons (toBuiltinList m)
+    (k, a) = P.pairToPair hd
+
+{-# INLINEABLE noDuplicateKeys #-}
+noDuplicateKeys :: forall k a. Map k a -> Bool
+noDuplicateKeys m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let k = BI.fst hd
+             in if member' k tl then False else go tl
+        )
+
+{-# INLINEABLE all #-}
+all :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
```

Comment:
why not define a fold and use that for a bunch of these?

Reply:
generally it seems like you've avoided defining any higher-order functions and using them, which seems like overkill to me, and makes this much harder to review...

Reply:
> why not define a fold and use that for a bunch of these?

That would make it trickier to make sure that things short-circuit properly, although inlining probably is more or less reliable in that case, although relying on inlining is always dangerous.

Reply:
> why not define a fold

Then `fold` will have to be inlined and beta-reduced, but there's no guarantee, even if it has the INLINE pragma. Recall that the INLINE pragma only makes it inline, not beta-reduce.

The code duplication shouldn't be a big deal due to CSE.